README.md
Description
===============

**ReleaseNoteMaker:** A PHP script to generate a complete Release Note list from your repository, starting from the first commit. It summarises the commits for each version number (defined by tags starting with a "v"). 

## Usage
 
    cd <path/to/repo>
    <path to ReleaseNotemaker.php> > ReleaseNotes.md

or, if you have commits that are not part of the latest tagged version, you can also pass on two arguments. If you don't provide at least the version number, then the script will try to estimate the upcoming version number – see below.

    cd <path/to/repo>
    <path to ReleaseNotemaker.php> > ReleaseNotes.md <versionsnummer> <shortdescription>


You probably want to clean up unnecessary or double commits in the final list.

The [CHANGELOG](CHANGELOG.md) of this project is created with the script.

### Filtering
The script contains a simple routine to filter out some commits. Per default, the following commits are **not** part of the release notes:

* Commits that contain `(minor)`
* Commits that contain `Todo aktualisiert`
* Commits that contain `(-)`

This can be changed in the script. You can add, delete or edit filters here:

```
$greppers=array("\(minor\)","Todo aktualisiert");
```

### Version number proposal
If there are commits that are not yet part of a release, the script will show them at the end of the log. It will also try to estimate the upcoming version number based on the number of commits and their commit messages:

- Commits with `(minor)` or other filtered strings in the commit message will be excluded from calculation (see above, *Filtering*).
- Commits with `(!)` in their message will be considered commits that **break compatibility**. This will always create a **major** release version number (e.g. from `v4.2.1` to `v5.0`).
- Commits with `(+)` will be considered commits that **add an important new feature**. This will always create a **minor** release version number (e.g. from `v5.0.2` to `v5.1`). Its commit text will be proposed as tag annotation.
- More than 10 normal commits will also create a **minor** release version number (e.g. from `v5.0.2` to `v5.1`). The proposal for the tag annotation will be `Maintenance release`. 
- Everything else (no (!) or (+) in the commit message, less than 10 commits) will be considered a bug fix release version number (e.g. from `v5.1` to `v5.1.1`). The proposal for the tag annotation will be `Small bugfix release`.

### Good practices

* Create a version tag starting with `v`, e.g. `v1.3.4`
* Add a short summary in the tag message, e.g. `Improved routine to calculate the entries properly. Fixes #2 and #3`
* Add `(!)` in the commit message when your commit introduces a major change in the functionality that would break compatibility or changes the way the user would work with the software. E.g. if you change the database format or the way you store information in a file. This will automatically lead to a *major release*.
* Add a `(+)` to your commit message if you introduce a new function. This will automatically lead to a *minor release*.
* Use the filter function to hide commits from the logfile. E.g. use `(minor)` in the commit message when you just fixed a typo, or if you cleaned up the code etc.
* Adjust the filters according to your workflow and preferences 

Contact
===============
arthurbauer@mac.com

CHANGELOG.md
# Release notes
## v0.1
**First release**  
2015-12-16  
*Changes from `project start` to `v0.1`:*

* Created empty script
* First running version of the script
* Added a small routine to create release notes for the first version as well.
* Updated version number

## v0.2
**Second release**  
2015-12-16  
*Changes from `v0.1` to `v0.2`:*

* Updated the Markdown output a bit
* Bugfix: Added output for latest changes again
* F001: Rename first and last commit in output - FIXED
* Added header and changed access rights to make the script executable
* Adjusted Markdown output
* Removed path to project from Release notes header
* Improved markdown for subline
* Release notes and version number updated

## v0.3
**Third release**  
2015-12-17  
*Changes from `v0.2` to `v0.3`:*

* Added a "--reverse"  to bring commits in a more logical order
* Adjusted some comments in the source code
* Improved output for latest, not yet released changes
* Only creates output if there is actually anything new since the last version tag
* Bugfix: Last entry was repeated under certain circumstances
* Cleaner sorting of versions
* Version number updated

## v0.4
**Fourth release**  
2015-12-17  
*Changes from `v0.3` to `v0.4`:*

* Switched to a more reliable sort function for version-numbers
* Version number updated

## v0.5
**Last development release**  
2015-12-18  
*Changes from `v0.4` to `v0.5`:*

* Added tag message to output
* Small bugfix for tag commit message that messed up the markdown
* New routine to add the commit date to the notes
* Bugfix on the date routine
* Version number updated

## v1.0
**First release**  
2016-01-11  
*Changes from `v0.5` to `v1.0`:*

* Changed grep routine to make it easier to add or remove grep strings
* Added possibility to add parameters
* Version number updated

## v1.1
**Bugfix release**  
2016-01-25  
*Changes from `v1.0` to `v1.1`:*

* If no tags found, get the first commit instead. Closes #1
* Add description about commit filter. Closes #2
* Updated readme and version number

## v1.2
**Version number proposal**  
2016-02-22  
*Changes from `v1.1` to `v1.2`:*

* Updated documentation
* Reformat whitespace. Added Coda style comments
* Version number proposal (+)
* Little bug fix if there was no tag yet
* Updated README.md, added version number proposal explanation.
* Updated README.md to explain the new syntax for commit messages.

## v1.2.1
**Small bugfix release**  
2016-03-09  
*Changes from `v1.2` to `v1.2.1`:*

* Little bug fix if there was no tag yet
* Changed grep routine to be case-sensitive
* Added a routine to filter out duplicate commits


ReleaseNoteMaker.php
#!/usr/bin/php
<?php
// v1.2.1 2016-03-09

/* CONFIG AREA */

$greppers=array("\(minor\)", "Todo aktualisiert","(-)"); // all these strings are filtered out
$tagmajor="(!)"; // this string creates a major release
$tagminor="(+)"; // this string creates a minor release

/* END CONFIG AREA */

date_default_timezone_set("Europe/Berlin");

$cv=$argv;

$tags=trim(`git tag -l "v*" `); // get all the tags that start with a "v"

$tags=explode("\n", $tags);

// a short routine to sort prerelease versions (with different type of suffixes) before their full version
// it is important that one follows the semantic versioning principle closely:
//
//  vx.z.y-prerelease OR vx.z.y_prerelease
//
// dash and underscore should not be mixed
//

usort($tags, 'version_compare');

if (!$tags[0]) $notags=true;

$counter=count($tags)-1;

$tags[$counter+1]="HEAD"; // and the current HEAD as last "version"

//start output
echo "# Release notes";
//echo `pwd`;

//! Get the first commit and add it at the beginning of the list, so we also get the first commit messages
$commit1=`git log --oneline | tail -n 1`;
$commit1=explode(" ", $commit1);
array_unshift($tags, $commit1[0]);

if ($notags)
{
	$tags[0]=trim(`git log --reverse  --oneline | head -n 1 | awk '{print $1}'`);
	$tags[1]="HEAD";
	$firstcom=$tags[0];
}

//! loop through all the tags and create an output for that version
for ($i=0;$i<=$counter+1;$i++)
{
	$j=$i+1;

	//! This command creates the actual release note, greps out commits that contain "(minor)", "Todo" and "Version number"/"Versionsnummer"
	// Feel free to add more greps if needed

	$grepstring="";

	foreach ($greppers as $grep1) $grepstring.=" | grep -v '$grep1'";

	$com="git log --reverse --no-merges --pretty=format:\"* %s\" ".$tags[$i]."..".$tags[$j]."  ".$grepstring;

	if ($tags[$j]!="HEAD")
	{
		$log= "\n## $tags[$j]\n";

		//! let's get the tag message here and add it to the output
		$tagmessage=`git show $tags[$j]`;
		$tagmessage=explode("\n", $tagmessage);
		if ($tagmessage[4]) $log.="**".trim($tagmessage[4])."**  ";

		//! let's get the commit date here and add it to the output
		$tagtime=`git log --pretty=tformat:%at $tags[$j] | head -n 1`;
		if ($tagtime) $log.="\n".date("Y-m-d", $tagtime)."  \n";



		$log.= "*Changes from `$tags[$i]` to `$tags[$j]`:*\n\n";
		$log.= join("\n",array_unique(explode("\n",`$com`)));
	}
	else
	{
		//! let's check if there is anything new anyway
		$check=`$com`;
		if ($check!==NULL)
		{
			if (!$cv[1])
			{
				//! here we have a closer look at the new stuff to find out what kind of version number we get.
				$closelook=explode("\n",trim($check));
				$closelookcount=count($closelook);
				$major=$minor=$bugfix=false;
				foreach ($closelook as $closelookline)
				{
				if (strpos($closelookline, $tagmajor))
				{
					$major=true;
					$annot[]=trim(str_replace("* ","",trim(str_replace($tagmajor,"",$closelookline))));
				}
				if (strpos($closelookline, $tagminor)) 
				{
					$minor=true;
					$annot[]=trim(str_replace("* ","",trim(str_replace($tagminor,"",$closelookline))));
				
				}
				}
		
				$oldvn=explode(".",str_replace("v","",$tags[$i]));
				if ($major) 
				{
					$newvn=($oldvn[0]+1).".0";
					if (!$annot) $annot[]="Major release";
				}
				else if ($minor or $closelookcount>10) {
					$newvn=$oldvn[0].".".($oldvn[1]+1);
					if (!$annot) $annot[]="Maintenance release";
					//else $annot[]="feature release";
				}
				else 
				{
					$newvn=$oldvn[0].".".$oldvn[1].".".($oldvn[2]+1);
					$annot[]="Small bugfix release";
}
				$log= "\n## Upcoming version v$newvn *(not yet released)*  \n";
				if ($annot) $log.="**New:** ".join(", ",$annot)."  \n";
				$log.=date("Y-m-d")."  \n";
				$log.= "*Changes from `$tags[$i]` to current version:*\n\n";
			}
			else
			{
				$log= "\n## $cv[1]".($cv[2]?"\n**$cv[2]**  ":"");
				$log.="\n".date("Y-m-d")."  \n";
				$log.= "*Changes from `$tags[$i]` to `$cv[1]`:*\n\n";
			}
			$log.= join("\n",array_unique(explode("\n",`$com`)));
		}
	}


	//! replace "HEAD" and the first commit hash with a more human-readable text.
	$log=str_replace("HEAD", "current version", $log);
	$log=str_replace($commit1[0], "project start", $log);
	$log=str_replace("vproject start..1", "v0.1", $log);
	$log=str_replace("vproject start.1", "v0.1", $log);

	echo $log;
	$log="";
}
?>